### Identify what it is we want to do  / loose requirements
* Want to know where we can go camping near Mount Hood
* Want to know how many sites are at a campground
* Want to know what sites (number of sites?) are accessible, near water, toilet y/n, and allow pets

#### This looks like:  
A dataset we can query with criteria and get a list of campsites that match, ideally with info and weblinks for further research.

### Explore data to determine if / how we can do that  
* Identify attributes in RIDB with this info
* Join with data from forest service websites

### Productionalize for scale
* generic code for getting RIDB data, configured by JSON for specific locations. Store this in a db / lookup so we can onboard new campsites without a deploy
* parallelizing to reduce runtime - api rate limits
* how often should the pipeline run?
* replace or append?
* metadata - data source, ingested_on timestamp
* ERrors - just like you want to know when something isnt as expected at a campground


### What are some camping analogies that might relate to what Im trying to convey about scalability?
* Send friends to multiple campgrounds to help find sites instead of everyone going to the same place - parallelization
* Come back to rendevous point because no cell service and compare what we found - 5 sites near water, no accesable sites, 
* Ikea camping chair?

In [119]:
import json
import pandas as pd
import itertools
import requests

import config

pd.set_option('display.max_colwidth', None)

Getting a list of facilities from RIDB
lat/long  
or state as comma delimited list of 2 char state codes  
https://ridb.recreation.gov/docs  
https://ridb.recreation.gov/ridb-access-agreement 

In [13]:
ridb_facilities_url = "https://ridb.recreation.gov/api/v1/facilities"

In [197]:
# params = {"activity_id":9, "latitude":45.4977712, "longitude":-121.8211673, "radius":15}
params = {"activity_id":9, "state":"OR"}
headers = {"accept": "application/json", "apikey": "b2b7cf12-0a8a-413b-aa75-3125b41977b8"}
response = requests.get(ridb_facilities_url, params, headers=headers)
# with open('../data/RIDB_OR_facilities.json', 'w') as f:
#     json.dump(response.text, f)
camping_json  = json.loads(response.text)

In [199]:
with open('../data/RIDB_OR_facilities.json') as f:
    res = json.load(f)
    
res

'{"RECDATA":[{"FacilityID":"265224","LegacyFacilityID":"","OrgFacilityID":"77603","ParentOrgID":"131","ParentRecAreaID":"1108","FacilityName":"Glendale - Powers Bicycle Recreation Area","FacilityDescription":"\\u003cp\\u003eThe Oregon Coast is a cycling attraction that is gaining worldwide recognition for its stunning and diverse landscapes, and the areas just inland from that majestic Highway 101 are equally impressive! The Glendale-Powers route offers the opportunity to bicycle through the ancient forests\xa0on the west side of the\xa0Rogue River-Siskiyou National Forest, along the beautiful Coquille River! Escape into a world of pristine beauty as you pass through\xa0wildflower-filled meadows, ride along clean, clear waters, and enjoy the majesty of the towering trees. Every turn of the road reveals all manner of big and small wildlife, including elk, bobcat, quail, and trout. This area provides for such amazing road biking opportunities, there is even an family-friendly annual ride

In [184]:
response.status_code

200

In [189]:
response.reason

'OK'

In [190]:
camping_json.RECDATA

AttributeError: 'dict' object has no attribute 'RECDATA'

In [181]:
camping_json['RECDATA'][0].keys()

dict_keys(['FacilityID', 'LegacyFacilityID', 'OrgFacilityID', 'ParentOrgID', 'ParentRecAreaID', 'FacilityName', 'FacilityDescription', 'FacilityTypeDescription', 'FacilityUseFeeDescription', 'FacilityDirections', 'FacilityPhone', 'FacilityEmail', 'FacilityReservationURL', 'FacilityMapURL', 'FacilityAdaAccess', 'GEOJSON', 'FacilityLongitude', 'FacilityLatitude', 'Keywords', 'StayLimit', 'Reservable', 'Enabled', 'LastUpdatedDate'])

In [160]:
df_ridb_camping = pd.DataFrame(camping_json['RECDATA'])

In [161]:
df_ridb_camping.columns

Index(['FacilityID', 'LegacyFacilityID', 'OrgFacilityID', 'ParentOrgID',
       'ParentRecAreaID', 'FacilityName', 'FacilityDescription',
       'FacilityTypeDescription', 'FacilityUseFeeDescription',
       'FacilityDirections', 'FacilityPhone', 'FacilityEmail',
       'FacilityReservationURL', 'FacilityMapURL', 'FacilityAdaAccess',
       'GEOJSON', 'FacilityLongitude', 'FacilityLatitude', 'Keywords',
       'StayLimit', 'Reservable', 'Enabled', 'LastUpdatedDate'],
      dtype='object')

In [175]:
df_ridb_camping.query("FacilityID == '251434'")

FacilityID LegacyFacilityID OrgFacilityID ParentOrgID ParentRecAreaID  \
7     251434           125541      AN425541         131            1106   

                      FacilityName  \
7  LOST LAKE RESORT AND CAMPGROUND   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       FacilityDescription  \
7  <h2>Overview</h2>\n<p>Lost Lake Campground is couched between beautiful Lost Lake and Mount Hood.  From Lost Lake Butte, the valley view is breathtaking, and from the viewpoint, guests can revel in the most majestic view of Mt. Hood, its glaciers and Lost Lake.  Another one of the many attractions at Lost Lake Campground is the old growth boardwalk, an interpretive trail that meanders through one of Oregon's large old growth stands.</p>\n<h2>Recreation</h2>\n<p>Lost Lake Campground offers an abundance of activities.  Campers have access to a general store, which offers: canoe, kayak, historic wooden row boats, metal fishing boats and stand-up paddleboard rentals.   Whether you are a professional angler looking to catch that prized German Brown or a first-timer, there is something for everyone.  Additionally, there are a number of hiking trails near the campground ranging in length from 1/4 mile to over 100 miles resulting in numerous hiking and biking opportunities.  Other activities include bird watching, berry picking, swimming, waterfall finding and more.</p>\n<h2>Facilities</h2>\nLost Lake Campground is adjacent to a full service resort with a variety of campsites to accommodate all types of visitors, including group, family and single campsites.  In addition to the campsites, there are cabins, yurts, lodge rooms and an extensively stocked general store on site.  <br/> <br/> \n\nThere is drinking water at designated locations and pit toilets throughout.   Each site has a designated tenant

In [176]:
df_ridb_camping.query("FacilityID == '251434'")[['FacilityLongitude', 'FacilityLatitude']]

FacilityLongitude  FacilityLatitude
7        -121.821944         45.488889

For each facility we want to get campground information

In [80]:
# Hmm! Not all these facilities are campgrounds

In [164]:
df_ridb_camping[['FacilityID', 'FacilityName']]

FacilityID                             FacilityName
0      233329                 TILLY JANE GUARD STATION
1      232834                   RILEY HORSE CAMPGROUND
2      234306            EAGLE CREEK OVERLOOK GRP SITE
3      234075                 WILDWOOD RECREATION SITE
4      122940                         WYETH CAMPGROUND
5    10070260                    Vista Ridge Trailhead
6      272093                      SHERWOOD CAMPGROUND
7      251434          LOST LAKE RESORT AND CAMPGROUND
8      232836                                 TOLLGATE
9      232835                              STILL CREEK
10   10070571                   Herman Creek Trailhead
11     232838                               LOST CREEK
12     248545                         Timberline Lodge
13     273800        Dog Mountain Trail System Permits
14     122890                   EAGLE CREEK CAMPGROUND
15     252341                   Wahtum Lake Campground
16     252385                   Zigzag Ranger District
17     252367               Hood River Ranger District
18     252397                       Top Spur Trailhead
19     248300           Bradford Island Visitor Center
20     271477                         Mt. Hood Meadows
21     252416                    Mirror Lake Trailhead
22     252361                      Climbing Mount Hood
23     242418                 Eagle Creek Day Use Area
24     248302             Navigation Lock Visitor Area
25     248303  Washington Shore Visitor Center Complex
26     248301                         Home Valley Park
27     272092  Kinnikinnick (Laurance Lake) Campground
28     252368            Little Zigzag Falls Trailhead
29     252336                  Paradise Park Trailhead
30     232837                               CAMP CREEK
31     233326                       TILLY JANE A-FRAME

In [82]:
# Get campsite data for each area, if no campsites then drop --- or should we?
# Consider how we would scale this out - 
# Lost Lake example

In [192]:
f"{ridb_facilities_url}/251434/campsites".split("/")[-2]

'251434'

In [200]:
resp = requests.get(f"{ridb_facilities_url}/251434/campsites", headers=headers)

In [201]:
resp.status_code

200

In [202]:
with open('../data/RIDB/campsites/251434.json', 'w') as f:
    json.dump(resp.text, f)

In [146]:
campsites = json.loads(resp.text)

In [147]:
len(campsites['RECDATA'])

50

In [148]:
df_campsites = pd.DataFrame(campsites['RECDATA'])

In [149]:
df_campsites

CampsiteID FacilityID  CampsiteName                     CampsiteType  \
0       96115     251434          A021                   RV NONELECTRIC   
1       96043     251434          C005                   RV NONELECTRIC   
2       96075     251434          D001                   RV NONELECTRIC   
3       96104     251434          A009                   RV NONELECTRIC   
4       96032     251434          H001           EQUESTRIAN NONELECTRIC   
5       96014     251434          B012                   RV NONELECTRIC   
6       96127     251434          B010                   RV NONELECTRIC   
7       96083     251434          A017                   RV NONELECTRIC   
8       96023     251434          B004                   RV NONELECTRIC   
9       96040     251434          D012            TENT ONLY NONELECTRIC   
10      96156     251434          C002                   RV NONELECTRIC   
11      96103     251434          H003           EQUESTRIAN NONELECTRIC   
12      96065     251434          D016                   RV NONELECTRIC   
13      96149     251434          D009            TENT ONLY NONELECTRIC   
14      96003     251434          A022                   RV NONELECTRIC   
15      96099     251434          F004            TENT ONLY NONELECTRIC   
16      96107     251434          A004                   RV NONELECTRIC   
17      96086     251434          D018                   RV NONELECTRIC   
18      96067     251434          B003                   RV NONELECTRIC   
19      96278     251434          F005            TENT ONLY NONELECTRIC   
20      96176     251434   A007a and b  GROUP STANDARD AREA NONELECTRIC   
21      96033     251434          B006                   RV NONELECTRIC   
22      96140     251434          B009                   RV NONELECTRIC   
23      96224     251434          F006            TENT ONLY NONELECTRIC   
24      96101     251434          H005           EQUESTRIAN NONELECTRIC   
25      96212     251434  A006 a and b  GROUP STANDARD AREA NONELECTRIC   
26      96080     251434          D002                   RV NONELECTRIC   
27      96112     251434          D020                   RV NONELECTRIC   
28      96072     251434          D005                   RV NONELECTRIC   
29      96056     251434          B007                   RV NONELECTRIC   
30      96012     251434          A014                   RV NONELECTRIC   
31      96097     251434          D010            TENT ONLY NONELECTRIC   
32      96136     251434          D007                   RV NONELECTRIC   
33      96114     251434          D024                   RV NONELECTRIC   
34      96098     251434          A011                   RV NONELECTRIC   
35      96028     251434          B013            TENT ONLY NONELECTRIC   
36      96131     251434          C006                   RV NONELECTRIC   
37      96019     251434          C010                   RV NONELECTRIC   
38      96141     251434          A012                   RV NONELECTRIC   
39      96004     251434          A019                   RV NONELECTRIC   
40      96095     251434          A020                   RV NONELECTRIC   
41      96264     251434          F002            TENT ONLY NONELECTRIC   
42      96057     251434          B014            TENT ONLY NONELECTRIC   
43      96073     251434          D017                   RV NONELECTRIC   
44      96111     251434          C008                   RV NONELECTRIC   
45      96303     251434          F001            TENT ONLY NONELECTRIC   
46      96053     251434          B011                   RV NONELECTRIC   
47      96013     251434          B002                   RV NONELECTRIC   
48      96009     251434          D004                   RV NONELECTRIC   
49      96085     251434          D015                   RV NONELECTRIC   

    TypeOfUse       Loop  CampsiteAccessible  CampsiteLongitude  \
0   Overnight     A Loop               False        -121.816561   
1   Overnight     C Loop                Tru

50 campsites, compare with the lostlake website that says 125:  
https://www.fs.usda.gov/recarea/mthood/recarea/?recid=53228 45.48889, -121.82194

In [150]:
[entry for entry in df_campsites.iloc[0].ATTRIBUTES]

[{'AttributeName': 'Driveway Grade', 'AttributeValue': 'Slight'},
 {'AttributeName': 'Max Num of Vehicles', 'AttributeValue': '1'},
 {'AttributeName': 'Fire Pit', 'AttributeValue': 'Y'},
 {'AttributeName': 'Privacy', 'AttributeValue': 'Y'},
 {'AttributeName': 'Base Number of Vehicles', 'AttributeValue': '0'},
 {'AttributeName': 'Base Number of People', 'AttributeValue': '0'},
 {'AttributeName': 'Grills/Fire Ring', 'AttributeValue': 'Y'},
 {'AttributeName': 'Placed on Map', 'AttributeValue': '1'},
 {'AttributeName': 'Picnic Table', 'AttributeValue': 'Y'},
 {'AttributeName': 'Checkin Time', 'AttributeValue': '3:00 PM'},
 {'AttributeName': 'Location Rating', 'AttributeValue': 'Good'},
 {'AttributeName': 'Site Rating', 'AttributeValue': 'Preferred'},
 {'AttributeName': 'Capacity/Size Rating', 'AttributeValue': 'Single'},
 {'AttributeName': 'Checkout Time', 'AttributeValue': '10:00 AM'},
 {'AttributeName': 'Max Num of People', 'AttributeValue': '5'},
 {'AttributeName': 'Campfire Allowed', '

Do the campsite attributes have the information we are looking for? Near water, accessible...

In [151]:
set(itertools.chain(*df_campsites['ATTRIBUTES'].apply(lambda x: [entry['AttributeName'] for entry in x])))

{'Accessibility',
 'BBQ',
 'Base Number of People',
 'Base Number of Vehicles',
 'Campfire Allowed',
 'Capacity/Size Rating',
 'Checkin Time',
 'Checkout Time',
 'Double Driveway',
 'Driveway Entry',
 'Driveway Grade',
 'Driveway Length',
 'Driveway Surface',
 'Fire Pit',
 'Grills/Fire Ring',
 'Hike In Distance to Site',
 'Horse Hitching Post',
 'Horse Stall/Corral',
 'IS EQUIPMENT MANDATORY',
 'Location Rating',
 'Max Num of People',
 'Max Num of Vehicles',
 'Max Vehicle Length',
 'Min Num of People',
 'Pets Allowed',
 'Picnic Table',
 'Placed on Map',
 'Platform',
 'Privacy',
 'Proximity to Water',
 'Quiet Area',
 'Shade',
 'Site Access',
 'Site Height/Overhead Clearance',
 'Site Rating',
 'Tent Pad',
 'Tent Pad Length',
 'Tent Pad Width'}

near water: "Proximity to Water"   
pets allowed: "Pets Allowed"   
accessibility: "Accessibility" - boolean  
toilet?  - no info, but thats what the forest service sites have

In [152]:
df_campsites.to_json('../data/ridb_campsites_mock.json', orient='records')

In [153]:
with open('../data/ridb_campsites_mock.json') as f:
    ridb_json = json.loads(f.read())
    
df = pd.json_normalize(ridb_json, "ATTRIBUTES", ["CampsiteID", "FacilityID"])

In [154]:
df

AttributeName AttributeValue CampsiteID FacilityID
0              Driveway Grade         Slight      96115     251434
1         Max Num of Vehicles              1      96115     251434
2                    Fire Pit              Y      96115     251434
3                     Privacy              Y      96115     251434
4     Base Number of Vehicles              0      96115     251434
...                       ...            ...        ...        ...
1371         Campfire Allowed            Yes      96085     251434
1372      Max Num of Vehicles              1      96085     251434
1373         Driveway Surface          Paved      96085     251434
1374           Driveway Grade         Slight      96085     251434
1375       Max Vehicle Length             80      96085     251434

[1376 rows x 4 columns]

In [157]:
df.query("AttributeName == 'Pets Allowed'").head()

AttributeName  AttributeValue CampsiteID FacilityID
16   Pets Allowed             Yes      96115     251434
43   Pets Allowed             Yes      96043     251434
70   Pets Allowed             Yes      96075     251434
105  Pets Allowed             Yes      96104     251434
123  Pets Allowed  Domestic,Horse      96032     251434

In [159]:
df.query("AttributeName == 'Proximity to Water'")

AttributeName AttributeValue CampsiteID FacilityID
425   Proximity to Water      Lakefront      96099     251434
542   Proximity to Water      Lakefront      96278     251434
638   Proximity to Water      Lakefront      96224     251434
1152  Proximity to Water      Lakefront      96264     251434
1248  Proximity to Water      Lakefront      96303     251434

In [170]:
df_ridb_camping.dtypes

FacilityID                    object
LegacyFacilityID              object
OrgFacilityID                 object
ParentOrgID                   object
ParentRecAreaID               object
FacilityName                  object
FacilityDescription           object
FacilityTypeDescription       object
FacilityUseFeeDescription     object
FacilityDirections            object
FacilityPhone                 object
FacilityEmail                 object
FacilityReservationURL        object
FacilityMapURL                object
FacilityAdaAccess             object
GEOJSON                       object
FacilityLongitude            float64
FacilityLatitude             float64
Keywords                      object
StayLimit                     object
Reservable                      bool
Enabled                         bool
LastUpdatedDate               object
dtype: object

In [171]:
df.merge(df_ridb_camping, on='FacilityID', how='left')

AttributeName AttributeValue CampsiteID FacilityID  \
0              Driveway Grade         Slight      96115     251434   
1         Max Num of Vehicles              1      96115     251434   
2                    Fire Pit              Y      96115     251434   
3                     Privacy              Y      96115     251434   
4     Base Number of Vehicles              0      96115     251434   
...                       ...            ...        ...        ...   
1371         Campfire Allowed            Yes      96085     251434   
1372      Max Num of Vehicles              1      96085     251434   
1373         Driveway Surface          Paved      96085     251434   
1374           Driveway Grade         Slight      96085     251434   
1375       Max Vehicle Length             80      96085     251434   

     LegacyFacilityID OrgFacilityID ParentOrgID ParentRecAreaID  \
0              125541      AN425541         131            1106   
1              125541      AN425541         131            1106   
2              125541      AN425541         131            1106   
3              125541      AN425541         131            1106   
4              125541      AN425541         131            1106   
...               ...           ...         ...             ...   
1371           125541      AN425541         131            1106   
1372           125541      AN425541         131            1106   
1373           125541      AN425541         131            1106   
1374           125541      AN425541         131            1106   
1375           125541      AN425541         131            1106   

                         FacilityName  \
0     LOST LAKE RESORT AND CAMPGROUND   
1     LOST LAKE RESORT AND CAMPGROUND   
2     LOST LAKE RESORT AND CAMPGROUND   
3     LOST LAKE RESORT AND CAMPGROUND   
4     LOST LAKE RESORT AND CAMPGROUND   
...                               ...   
1371  LOST LAKE RESORT AND CAMPGROUND   
1372  LOST LAKE RESORT AND CAMPGROUND   
1373  LOST LAKE RESORT AND CAMPGROUND   
1374  LOST LAKE RESORT AND CAMPGROUND   
1375  LOST LAKE RESORT AND CAMPGROUND   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     